server/src/main/java/org/elasticsearch/action/search/SearchPhase.java

```diff
@@ -62,4 +65,35 @@ static void doCheckNoMissingShards(String phaseName, SearchRequest request, Grou
             }
         }
     }
+
+    /**
+     * Releases shard targets that are not used in the docsIdsToLoad.
+     */
+    static void releaseIrrelevantSearchContext(SearchPhaseResult searchPhaseResult, SearchPhaseContext context) {
```

Comment:
Please assert that if `searchPhaseResult` isn't null, that its only a rank or fetch phase result?

Reply:
I think that this should be either a `queryResult` (what we had up until now) or a `rankFeatureResult` (the new addition), and within the method we'e also added that
```
        SearchPhaseResult phaseResult = searchPhaseResult.queryResult() != null
            ? searchPhaseResult.queryResult()
            : searchPhaseResult.rankFeatureResult();
```
 Do you mean to add an XOR check for the two; to ensure that we don't somehow end up with a mixed state?

Reply:
`SearchPhaseContext` I think. Basically, this was only used in fetch before. Now we are allowing two things, but we should make sure its restricted only to those two.

Reply:
Currently, we proceed execution only if we have a `queryResult` or `rankFeatureResult` specified.  So, this should be enabled for all phases that operate on these results (in this case, everything after `RankFeaturePhase`).

Not sure how we can restrict access to just `RankFeaturePhase` and `FetchSearchPhase`  based just on the `context` (or `searchPhaseResult`). Maybe we could add the function to a new `abstract class ReleaseableSeachPhase extends SearchPhase` and make `RankFeaturePhase` and `FetchSearchPhase` extend it? (or just copying the method to `RankFeaturePhase` and have some code duplication)? 🤔